enginetest/harness.go

```diff
@@ -35,6 +40,14 @@ type Harness interface {
 	// additional information (e.g. current DB) set uniformly. To replicated the behavior of tests during setup,
 	// harnesses should generally dispatch to enginetest.NewContext(harness), rather than calling this method themselves.
 	NewContext() *sql.Context
+	// NewEngine creates a new engine
+	NewEngine(*sql.Context, *testing.T) *sqle.Engine
+}
+
+type CheckpointHarness interface {
+	Harness
+	// RestoreCheckpoint resets the database to a saved point
+	RestoreCheckpoint(*sql.Context, *testing.T, *sqle.Engine) *sqle.Engine
```

Comment:
Rather than this, make NewEngine the sole method in CheckpointHarness. Then in each test method, if the harness implements CheckpointHarness, get a new engine from that. Otherwise, call enginetests.NewEngine().

Then the lifecycle of these objects is clear: it does any initialization work (including data setup) at instantiation, and creates a new blank engine with that initial state every time this method is called. No need for explicit checkpoint restore operations.

Reply:
Also document this lifecycle in these interfaces/